### PR TITLE
[flutter_tool] add enable-embedder-api flag

### PR DESCRIPTION
**Pull request by cyanglaz**
This flag is to enable the embedder api on iOS during iOS embedder api migration. Framework part of https://github.com/flutter/engine/pull/40025

First step of https://github.com/flutter/flutter/issues/112232

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I signed the [CLA].
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making, or this PR is [test-exempt].
- [x] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[test-exempt]: https://github.com/flutter/flutter/wiki/Tree-hygiene#tests
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat

### CODE REVIEW

File: packages/flutter/lib/src/scheduler/binding.dart [hunk 1]
```diff
@@ -385,6 +385,7 @@ mixin SchedulerBinding on BindingBase {
   @protected
   @mustCallSuper
   void handleAppLifecycleStateChanged(AppLifecycleState state) {
+    print('flutter: handle life cycle changes $state');
```
Comment (jmagman) [inline packages/flutter/lib/src/scheduler/binding.dart:388]: Stray print 🙂 
```suggestion
```

File: packages/flutter/lib/src/services/binding.dart [hunk 2]
```diff
@@ -259,6 +259,7 @@ mixin ServicesBinding on BindingBase, SchedulerBinding {
   }
 
   Future<String?> _handleLifecycleMessage(String? message) async {
+    print('flutter: _handleLifecycleMessage message: $message ');
```
Comment (jmagman) [inline packages/flutter/lib/src/services/binding.dart:262]:
```suggestion
```

File: packages/flutter_tools/lib/src/reporting/custom_dimensions.dart [hunk 3]
```diff
@@ -127,6 +128,7 @@ class CustomDimensions {
   final int? hotEventReloadVMTimeInMs;  // cd 55
   final bool? commandRunEnableImpeller;  // cd 56
   final String? commandRunIOSInterfaceType; // cd 57
+  final bool? commandRunEnableEmbedderApi; // cd 58
```
Comment (jmagman) [inline packages/flutter_tools/lib/src/reporting/custom_dimensions.dart:131]: Do we really need to track this as usage data? I can't imagine more than one or two non-contributor customers ever running with this (as opposed to impeller).

File: packages/flutter_tools/lib/src/reporting/custom_dimensions.dart [hunk 4]
```diff
@@ -187,6 +189,7 @@ class CustomDimensions {
       if (hotEventReloadVMTimeInMs != null) cdKey(CustomDimensionsEnum.hotEventReloadVMTimeInMs): hotEventReloadVMTimeInMs.toString(),
       if (commandRunEnableImpeller != null) cdKey(CustomDimensionsEnum.commandRunEnableImpeller): commandRunEnableImpeller.toString(),
       if (commandRunIOSInterfaceType != null) cdKey(CustomDimensionsEnum.commandRunIOSInterfaceType): commandRunIOSInterfaceType.toString(),
+      if (commandRunEnableImpeller != null) cdKey(CustomDimensionsEnum.commandRunEnableEmbedderApi): commandRunEnableEmbedderApi.toString(),
```
Comment (jmagman) [inline packages/flutter_tools/lib/src/reporting/custom_dimensions.dart:192]:
```suggestion
      if (commandRunEnableEmbedderApi != null) cdKey(CustomDimensionsEnum.commandRunEnableEmbedderApi): commandRunEnableEmbedderApi.toString(),
```

File: packages/flutter_tools/lib/src/runner/flutter_command.dart [hunk 5]
```diff
@@ -1043,6 +1043,16 @@ abstract class FlutterCommand extends Command<void> {
     );
   }
 
+  void addEnableEmbedderApiFlag({required bool verboseHelp}) {
+    argParser.addFlag('enable-embedder-api',
+        negatable: false,
```
Comment (jmagman) [inline packages/flutter_tools/lib/src/runner/flutter_command.dart:1048]: This should be negatable and default to `false`.

File: packages/flutter_tools/lib/src/runner/flutter_command.dart [hunk 6]
```diff
@@ -1043,6 +1043,16 @@ abstract class FlutterCommand extends Command<void> {
...
+        hide: !verboseHelp,
+        help: 'Whether to enable embedder api.'
+              'This flag only currently supports iOS, it is'
+              'ignored when targeting other platforms.',
```
Comment (jmagman) [inline packages/flutter_tools/lib/src/runner/flutter_command.dart:1052]: How about:
```suggestion
        help: 'Whether to enable the experimental embedder API on iOS.',
```

**Review by jmagman — CHANGES_REQUESTED** [review]

File: packages/flutter_tools/lib/src/reporting/custom_dimensions.dart (hunk 3 above)
Comment (cyanglaz) [reply packages/flutter_tools/lib/src/reporting/custom_dimensions.dart:131]: I will remove it.

**Review by jmagman — APPROVED** [review]
LGTM

**Pull request merged**